craft_store/base_client.py

```diff
@@ -31,6 +32,54 @@
 
 logger = logging.getLogger(__name__)
 
+# The known types of credentials: a string for Candid and a Dict for UbuntuOne
+CandidCredType = str
+UbuntuOneCredType = typing.Dict[str, str]
+CredentialsType = typing.Union[CandidCredType, UbuntuOneCredType]
+
+
+def wrap_credentials(token_type: str, credentials: CredentialsType) -> str:
+    """Create a payload string that contains both `credentials` and its identifying type.
+
+    This function creates a string that contains the desired `credentials` but also
+    stores their "type", for unwrapping later with `unwrap_credentials()`.
+
+    :param token_type: The identifier for the type of token stored in `credentials`
+    :param credentials: The actual type-dependent credentials.
+    :return: A payload string ready to be passed to Auth.set_credentials()
+    """
+    return json.dumps({"t": token_type, "v": credentials})
+
+
+def unwrap_credentials(
+    token_type: str, stored_credentials: str, deserialize_old_creds: bool
+) -> CredentialsType:
+    """Retrieve the type-specific "inner" credentials from `credentials`.
+
+    This function also handles backwards-compatibility by supporting `credentials`
+    from before we stored the `token_type`; it is meant to be called with the
+    returned value from Auth.get_credentials().
+
+    :param token_type: The expected token type, for validation.
+    :param stored_credentials: The credentials retrieved from auth storage.
+    :param deserialize_old_creds:
+        How to handle credentials stored in the old scheme: should they be deserialized
+        with JSON (True) or returned as-is (False).
+    """
+    try:
+        loaded = json.loads(stored_credentials)
+    except json.JSONDecodeError:
+        # Not a JSON string, so this is already the desired value
+        return stored_credentials
+
+    if "t" in loaded:
+        if loaded["t"] == token_type:
+            return loaded["v"]
+        raise errors.CredentialsNotParseable()
```

Comment:
Maybe we could define credentials as a pydantic model to make the implementation more uniform/robust, and easier to read in the future. We support multiple formats and testing by hand to support them all is confusing.

Reply:
good idea, thanks